File: webapp/storyteller/src/app/service/cover.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { CoverURI } from '../model/cover';

@Injectable({
  providedIn: 'root'
})

/**
 * Service for Cover
 */
export class CoverService {

  public visible : BehaviorSubject<boolean> = new BehaviorSubject(false);

  constructor( private http: HttpClient) { }

  /**
   * Get the cover preview
   * @param file the file to get the preview for
   */
  public previewCover(file : File) : Observable<CoverURI>{
    var fd = new FormData();
    fd.append('file', file);
    return this.http.post<CoverURI>('/api/v1/cover/previewImage', fd);
  }

  /**
   * Get the default cover uri
   */
  getDefaultUri(): CoverURI {
    let coverUri : CoverURI = {smImageUri: "assets/images/book-cover.png", originalImageUri: "assets/images/book-cover.png", lgImageUri: "assets/images/book-cover.png"} as CoverURI
    return coverUri; 
  }

  /**
   * Commit the image to the server
   * @param file the file to commit
   * @param bookId the book id to commit the image to
   */
  public commitImage(file : File, bookId : string) : Observable<CoverURI>{
    var fd = new FormData();
    fd.append('file', file);
    fd.append('bookId', bookId);
    return this.http.post<CoverURI>(`/api/v1/cover/commitImage`, fd);
  }

  /**
   * Show the cover
   */
  public show() {
    if (!this.visible.getValue()) {
      this.visible.next(true);
    }
  }

  /**
   * Hide the cover
   */
  public hide() {
    if (this.visible.getValue()) {
      this.visible.next(false);
    }
  }
}

File: api/src/main/java/de/storyteller/api/service/cover/CoverService.java
package de.storyteller.api.service.cover;


import de.storyteller.api.v1.dto.cover.CoverUriDTO;
import java.util.Optional;
import org.springframework.web.multipart.MultipartFile;

/**
 * Service for handling cover images.
 */
public interface CoverService {

    /**
     * Save a cover image for a book.
     * @param multipartFile The image file.
     * @param bookId The id of the book.
     * @return The uri of the saved image.
     */
    Optional<CoverUriDTO> save(MultipartFile multipartFile, String bookId);

    /**
     * Save a cover image for a book as preview.
     * @param multipartFile The image file.
     * @return The uri of the saved image.
     */
    Optional<CoverUriDTO> savePreview(MultipartFile multipartFile);
}
